### ISSUE TITLE
[Query] Common ORM like functionalities

### ISSUE THREAD

**Issue by sushilkjaiswar** [roman-right/beanie#135]
There are ORM functionality which I don't see in beanie which are mentioned below,
1. How to add default fields in the model which is updated from whenever  records are added. eg. created and updated fields.
2. I want certain fields to be added while insertion and certain fields only when record is updated. eg. created must be added value only at insertion of the records but updated field must be always updated whenever records is updated.

Is the above feature already present, if yes can you point me to the documentation.

**Comment by rdfsx** [comment]
I would like to know too

**Comment by roman-right** [comment]
Hey, sorry for the late reply.

I think you can do the thing with the next code:

```python
from datetime import datetime
from typing import Optional

from pydantic import Field

from beanie import Document, before_event, Replace, Insert


class Sample(Document):
    created_at: datetime = Field(default_factory=datetime.utcnow)
    updated_at: Optional[datetime] = None

    @before_event([Insert, Replace])
    def set_updated_at(self):
        self.updated_at = datetime.utcnow()

```

The `updated_at` field will be updated only before inserts and replaces here (and before `save()`, as it is a combination of `replace` and `insert` methods). But probably this is enough.

The doc about `before_event` could be found here: https://roman-right.github.io/beanie/tutorial/event-based-actions/

**Comment by sushilkjaiswar** [comment]
Thank you, Roman

I am referring to the document link you have posted but some how I missed this detail.

**Comment by rdfsx** [comment]
> Hey, sorry for the late reply.
> 
> I think you can do the thing with the next code:
> 
> ```python
> from datetime import datetime
> from typing import Optional
> 
> from pydantic import Field
> 
> from beanie import Document, before_event, Replace, Insert
> 
> 
> class Sample(Document):
>     created_at: datetime = Field(default_factory=datetime.utcnow)
>     updated_at: Optional[datetime] = None
> 
>     @before_event([Insert, Replace])
>     def set_updated_at(self):
>         self.updated_at = datetime.utcnow()
> ```
> 
> The `updated_at` field will be updated only before inserts and replaces here (and before `save()`, as it is a combination of `replace` and `insert` methods). But probably this is enough.
> 
> The doc about `before_event` could be found here: https://roman-right.github.io/beanie/tutorial/event-based-actions/

Can I use the Sample class as a mixin? The code below does not work (updated_at in ChatModel does not update)

`
class TimeBaseModel(Document):
    created_at: datetime = Field(default_factory=datetime.utcnow)
    updated_at: Optional[datetime] = None

    @before_event([Insert, Replace, SaveChanges])
    def set_updated_at(self):
        print("updated")
        self.updated_at = datetime.utcnow()


class ChatModel(TimeBaseModel):
    id: int = Field(alias="_id")
    type: str = Field(...)

    class Collection:
        name = "ChatModel"`

**Comment by roman-right** [comment]
Hey @rdfsx ,
Unfortunately, nope. I will fix this, but for now, it works only directly with the classes. It should be fixed in a week or two

**Comment by rdfsx** [comment]
> Hey @rdfsx , Unfortunately, nope. I will fix this, but for now, it works only directly with the classes. It should be fixed in a week or two

Okay, thank you for your work!

**Comment by roman-right** [comment]
Hey @rdfsx ,
It was fixed in 1.8.3 - you can inherit documents with actions now.

**Comment by rdfsx** [comment]
> Hey @rdfsx ,
> It was fixed in 1.8.3 - you can inherit documents with actions now.

Wow. Thank you so much!
